rcl_action/src/rcl_action/graph.c

```diff
@@ -0,0 +1,219 @@
+// Copyright 2019 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifdef __cplusplus
+extern "C"
+{
+#endif
+
+#include <assert.h>
+#include <string.h>
+
+#include "rcl/error_handling.h"
+#include "rcl/graph.h"
+#include "rcl/node.h"
+#include "rcutils/strdup.h"
+
+#include "rcl_action/graph.h"
+
+static
+rcl_ret_t
+_filter_action_names(
+  rcl_names_and_types_t * topic_names_and_types,
+  rcl_allocator_t * allocator,
+  rcl_names_and_types_t * action_names_and_types)
+{
+  assert(topic_names_and_types);
+  assert(allocator);
+  assert(action_names_and_types);
+
+  // Assumption: actions provide a topic name with the suffix "/_action/feedback"
+  // and it has type with the suffix "_FeedbackMessage"
+  const char * action_name_identifier = "/_action/feedback";
+  const char * action_type_identifier = "_FeedbackMessage";
+
+  rcl_ret_t ret;
+  const size_t num_names = topic_names_and_types->names.size;
+  char ** names = topic_names_and_types->names.data;
+
+  // Count number of actions to determine how much memory to allocate
+  size_t num_actions = 0u;
+  for (size_t i = 0u; i < num_names; ++i) {
+    const char * identifier_index = strstr(names[i], action_name_identifier);
```

Comment:
Should this actually check if `action_name_identifier` is a suffix and not anywhere in the middle of the string as the command above suggests?

Reply:
Yes, that would be stronger, but based on this comment, https://github.com/ros2/rcl/pull/411#discussion_r274555643, I thought it sufficient to leave this as-is. But, if you have a recommendation for a nice way to ensure it is at the end of the string I'm happy to add it.

Reply:
> if you have a recommendation for a nice way to ensure it is at the end of the string I'm happy to add it.

`identifier_index` needs to be `strlen(names[i]) - strlen(action_name_identifier)`?

Reply:
Since `strstr` returns a reference to the start of the substring, I think we want `strlen(identifier_index) == strlen(action_name_identifier)`, but seems like it's broken some tests. Looking into it.

Reply:
I think this does it: 661cab0